### PR TITLE
Various things

### PR DESCRIPTION
**Pull request by hmaurer**

### CODE REVIEW

File: src/smallstack.rs [hunk 1]
```diff
@@ -207,26 +207,28 @@ impl<'a> VM<'a> {
 }
 
 fn main() {
+    let mut content = String::new();
```
Comment (vendethiel) [inline src/smallstack.rs:208]: Indentation

File: src/smallstack.rs [hunk 2]
```diff
@@ -207,26 +207,28 @@ impl<'a> VM<'a> {
...
-      }
-    }
-	} else {
-    panic!("Please give a file argument");
-  }
+    } else {
+        result = std::io::stdin().read_to_string(&mut content);
+   }
+
+   if let Ok(_) = result {
```
Comment (vendethiel) [inline src/smallstack.rs:224]: try!() ?

Comment (hmaurer) [inline src/smallstack.rs:224]: try!() returns from the function on error, no? and it expects the return type to be Result<>

File: src/smallstack.rs (hunk 1 above)
Comment (hmaurer) [inline src/smallstack.rs:208]: What's your convention? Indentation in the whole file is a bit dodgy. Two white spaces? Four? Tabs?

File: src/smallstack.rs (hunk 2 above)
Comment (vendethiel) [inline src/smallstack.rs:224]: sure. do we want to continue on error?

File: src/smallstack.rs [hunk 3]
```diff
@@ -92,33 +92,31 @@ impl<'a> VM<'a> {
...
+          println!("{}", arg);
         } else {
           panic!("VM error: incorrect arguments to `say`");
         },
 
+        &["call", "primitive", "concat"] => if let
+            (Some(Expr::Str(arg1)), Some(Expr::Str(arg2)))
+            =
+            (stack.pop(), stack.pop()) {
+          stack.push(Expr::Str(arg1 + &arg2));
```
Comment (vendethiel) [inline src/smallstack.rs:104]: somewhat feeling uneasy about this.

i guess we should have `&["call", "primitive", x] =>` and then try to find the function in some map (the stdlib).

how do you feel about this?

this'd mean passing a mut ref to the stack, I guess. Or we could have a struct encoding the arity. But then, how to store functions with different arities? Or do we pass a vec? But then we'll have to pattern match it again inside the function...

i've been thinking about this since the beginning. Not too sure how I want to go about it. But I feel this it is time to tackle it now that we add more functions

Comment (hmaurer) [inline src/smallstack.rs:104]: Totally agree

Comment (hmaurer) [inline src/smallstack.rs:104]: For now this doesn't make a big difference, though

Comment (vendethiel) [inline src/smallstack.rs:104]: Are you interested in tackling this?

Comment (vendethiel) [inline src/smallstack.rs:104]: You could be ze one to bring a 2nd file to smallstacj

File: src/smallstack.rs (hunk 1 above)
Comment (vendethiel) [inline src/smallstack.rs:208]: I think it's mostly 2 spaces atm

File: src/smallstack.rs (hunk 3 above)
Comment (hmaurer) [inline src/smallstack.rs:104]: I am not sure we should tackle this as a global hash of primitive functions. Maybe we could first tackle modules (as hashes of functions), then introduce the notion of a "native module" which would be implemented in Rust within the VM?

Comment (vendethiel) [inline src/smallstack.rs:104]: Those are orthogonal tho

File: src/smallstack.rs (hunk 2 above)
Comment (hmaurer) [inline src/smallstack.rs:224]: No, but the return value of the function main() is () (which I assume is Unit?), not Result<>, so it doesn't compile.

Comment (vendethiel) [inline src/smallstack.rs:224]: Ah troo.

File: src/smallstack.rs [hunk 4]
```diff
@@ -207,26 +205,28 @@ impl<'a> VM<'a> {
 }
 
 fn main() {
+    let mut content = String::new();
+    let result;
+
 	if let Some(path) = env::args().nth(1) {
```
Comment (vendethiel) [inline src/smallstack.rs:211]: do you mind fixing this

**Comment by vendethiel** [comment]
gonna merge this.

the stdlib can come later
